File: app/schemas/rag_knowledge.py
from typing import List, Optional

from pydantic import BaseModel, Field


class KnowledgeBaseCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(default=None, max_length=512)
    scope_type: str = Field(default="personal", pattern="^(personal|public|class|course)$")
    scope_id: Optional[int] = None
    chunk_strategy: str = Field(default="paragraph", pattern="^(paragraph|fixed)$")
    chunk_size: int = Field(default=700, ge=200, le=2000)
    chunk_overlap: int = Field(default=100, ge=0, le=500)
    embedding_model: str = Field(default="moka-ai/m3e-base", max_length=100)
    vector_store: str = Field(default="milvus", pattern="^(milvus)$")
    similarity_metric: str = Field(default="COSINE", pattern="^(COSINE|IP|L2)$")
    retrieval_mode: str = Field(default="hybrid", pattern="^(vector|keyword|hybrid)$")
    default_top_k: int = Field(default=5, ge=1, le=20)
    default_min_score: int = Field(default=45, ge=0, le=100)
    vector_weight: int = Field(default=62, ge=0, le=100)
    bm25_weight: int = Field(default=28, ge=0, le=100)
    title_weight: int = Field(default=10, ge=0, le=100)
    core_weight: int = Field(default=35, ge=0, le=100)


class KnowledgeBaseUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    description: Optional[str] = Field(default=None, max_length=512)
    scope_type: Optional[str] = Field(default=None, pattern="^(personal|public|class|course)$")
    scope_id: Optional[int] = None
    status: Optional[int] = Field(default=None, ge=0, le=1)
    chunk_strategy: Optional[str] = Field(default=None, pattern="^(paragraph|fixed)$")
    chunk_size: Optional[int] = Field(default=None, ge=200, le=2000)
    chunk_overlap: Optional[int] = Field(default=None, ge=0, le=500)
    embedding_model: Optional[str] = Field(default=None, max_length=100)
    vector_store: Optional[str] = Field(default=None, pattern="^(milvus)$")
    similarity_metric: Optional[str] = Field(default=None, pattern="^(COSINE|IP|L2)$")
    retrieval_mode: Optional[str] = Field(default=None, pattern="^(vector|keyword|hybrid)$")
    default_top_k: Optional[int] = Field(default=None, ge=1, le=20)
    default_min_score: Optional[int] = Field(default=None, ge=0, le=100)
    vector_weight: Optional[int] = Field(default=None, ge=0, le=100)
    bm25_weight: Optional[int] = Field(default=None, ge=0, le=100)
    title_weight: Optional[int] = Field(default=None, ge=0, le=100)
    core_weight: Optional[int] = Field(default=None, ge=0, le=100)


class TextImportRequest(BaseModel):
    kb_id: int
    title: str = Field(..., min_length=1, max_length=200)
    text: str = Field(..., min_length=1)


class PathImportRequest(BaseModel):
    kb_id: int
    path: str = Field(..., min_length=1, max_length=500)
    title: Optional[str] = Field(default=None, max_length=200)


class KnowledgeSearchRequest(BaseModel):
    question: str = Field(..., min_length=1, max_length=1000)
    kb_ids: Optional[List[int]] = None
    top_k: int = Field(default=5, ge=1, le=20)
    min_score: float = Field(default=0.0, ge=0.0, le=1.0)


class KnowledgeAskRequest(KnowledgeSearchRequest):
    session_id: Optional[str] = Field(default=None, max_length=64)


class KnowledgeBaseOut(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    owner_id: int
    scope_type: str
    scope_id: Optional[int] = None
    status: int
    document_count: int
    chunk_count: int
    created_at: str
    updated_at: str


class DocumentOut(BaseModel):
    id: int
    kb_id: int
    title: str
    source_type: str
    file_name: Optional[str] = None
    file_ext: Optional[str] = None
    status: str
    error_message: Optional[str] = None
    chunk_count: int
    char_count: int
    created_at: str
    updated_at: str

File: frontend/src/api/rag-knowledge.ts
import request from '@/utils/request'

export interface KnowledgeBase {
  id: number
  name: string
  description?: string
  owner_id: number
  scope_type: 'personal' | 'public' | 'class' | 'course'
  scope_id?: number
  status: number
  document_count: number
  chunk_count: number
  chunk_strategy: 'paragraph' | 'fixed'
  chunk_size: number
  chunk_overlap: number
  embedding_model: string
  vector_store: string
  similarity_metric: 'COSINE' | 'IP' | 'L2'
  retrieval_mode: 'vector' | 'keyword' | 'hybrid'
  default_top_k: number
  default_min_score: number
  vector_weight: number
  bm25_weight: number
  title_weight: number
  core_weight: number
  eval_score?: number
  eval_recall?: number
  eval_precision?: number
  eval_f1?: number
  eval_hit?: number
  eval_mrr?: number
  eval_sample_count?: number
  evaluated_at?: string
  config?: Record<string, any>
  created_at: string
  updated_at: string
}

export interface KnowledgeDocument {
  id: number
  kb_id: number
  title: string
  source_type: 'text' | 'upload' | 'path'
  file_name?: string
  file_ext?: string
  status: string
  error_message?: string
  chunk_count: number
  char_count: number
  created_at: string
  updated_at: string
}

export interface KnowledgeSource {
  chunk_id: number
  document_id: number
  kb_id: number
  title: string
  chunk_no: number
  score: number
  vector_score?: number
  bm25_score?: number
  title_score?: number
  core_score?: number
  content: string
}

export interface KnowledgeAskResponse {
  question: string
  answer: string
  sources: KnowledgeSource[]
  retrieval?: Record<string, any>
}

export interface KnowledgeBasePayload {
  name: string
  description?: string
  scope_type?: string
  scope_id?: number
  chunk_strategy?: 'paragraph' | 'fixed'
  chunk_size?: number
  chunk_overlap?: number
  embedding_model?: string
  vector_store?: string
  similarity_metric?: 'COSINE' | 'IP' | 'L2'
  retrieval_mode?: 'vector' | 'keyword' | 'hybrid'
  default_top_k?: number
  default_min_score?: number
  vector_weight?: number
  bm25_weight?: number
  title_weight?: number
  core_weight?: number
}

export function listKnowledgeBases(params?: { keyword?: string; include_public?: boolean }) {
  return request.get<KnowledgeBase[]>('/rag/knowledge/bases', { params })
}

export function createKnowledgeBase(data: KnowledgeBasePayload) {
  return request.post<KnowledgeBase>('/rag/knowledge/bases', data)
}

export function updateKnowledgeBase(id: number, data: Partial<KnowledgeBasePayload & { status: number }>) {
  return request.put<KnowledgeBase>(`/rag/knowledge/bases/${id}`, data)
}

export function deleteKnowledgeBase(id: number) {
  return request.delete(`/rag/knowledge/bases/${id}`)
}

export function listKnowledgeDocuments(kbId: number) {
  return request.get<KnowledgeDocument[]>(`/rag/knowledge/bases/${kbId}/documents`)
}

export function getKnowledgeBaseDetail(kbId: number) {
  return request.get<Record<string, any>>(`/rag/knowledge/bases/${kbId}/detail`)
}

export function evaluateKnowledgeBase(kbId: number) {
  return request.post<Record<string, any>>(`/rag/knowledge/bases/${kbId}/evaluate`, {}, { timeout: 120000 })
}

export function importKnowledgeText(data: { kb_id: number; title: string; text: string }) {
  return request.post<KnowledgeDocument>('/rag/knowledge/documents/text', data, { timeout: 0 })
}

export function importKnowledgePath(data: { kb_id: number; title?: string; path: string }) {
  return request.post<KnowledgeDocument>('/rag/knowledge/documents/path', data, { timeout: 0 })
}

export function uploadKnowledgeDocument(kbId: number, file: File, title?: string) {
  const form = new FormData()
  form.append('kb_id', String(kbId))
  if (title) form.append('title', title)
  form.append('file', file)
  return request.post<KnowledgeDocument>('/rag/knowledge/documents/upload', form, {
    timeout: 0,
    headers: { 'Content-Type': 'multipart/form-data' },
  })
}

export function deleteKnowledgeDocument(id: number) {
  return request.delete(`/rag/knowledge/documents/${id}`)
}

export function askKnowledge(data: { question: string; kb_ids?: number[]; top_k?: number; min_score?: number }) {
  return request.post<KnowledgeAskResponse>('/rag/knowledge/ask', data, { timeout: 120000 })
}

export function knowledgeHealth() {
  return request.get('/rag/knowledge/health')
}
